crypto/fipsmodule/ec/ec_nistp.c

```diff
@@ -259,3 +259,47 @@ void ec_nistp_point_add(const ec_nistp_meth *ctx,
   cmovznz(z3, ctx->felem_num_limbs, z2nz, z1, z_out);
 }
 
+// Returns i-th bit of the scalar (zero or one).
+// The caller is responsible for making sure i is within bounds of the scalar. 
+static int16_t get_bit(const EC_SCALAR *in, size_t i) {
+// |in->words| is an array of BN_ULONGs which can be either 8 or 4 bytes long.
+#if defined(OPENSSL_64_BIT)
+  OPENSSL_STATIC_ASSERT(sizeof(BN_ULONG) == 8, bn_ulong_not_eight_bytes);
+  return (in->words[i >> 6] >> (i & 63)) & 1;
+#else
+  OPENSSL_STATIC_ASSERT(sizeof(BN_ULONG) == 4, bn_ulong_not_four_bytes);
+  return (in->words[i >> 5] >> (i & 31)) & 1;
+#endif
+}
+
```

Comment:
Are any extra precautions needed to avoid the C-compiler attempting to optimize this logic?
```
static int16_t get_bit(const EC_SCALAR *in, size_t i) {
// |in->words| is an array of BN_ULONGs which can be either 8 or 4 bytes long.
#if defined(OPENSSL_64_BIT)
  OPENSSL_STATIC_ASSERT(sizeof(BN_ULONG) == 8, bn_ulong_not_eight_bytes);
  return (value_barrier_u64(in->words[i >> 6]) >> (i & 63)) & 1;
#else
  OPENSSL_STATIC_ASSERT(sizeof(BN_ULONG) == 4, bn_ulong_not_four_bytes);
  return (value_barrier_u32(in->words[i >> 5]) >> (i & 31)) & 1;
#endif
}
```